app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt

```diff
@@ -246,30 +270,46 @@ class ConversationViewModel @Inject constructor(
 
     private suspend fun fromMessageModelToMessageContent(message: Message): MessageContent =
         when (val content = message.content) {
-            is Asset -> {
-                val assetId = content.value.remoteData.assetId
-                val (imgWidth, imgHeight) = when (val metadata = content.value.metadata) {
-                    is Image -> metadata.width to metadata.height
-                    else -> 0 to 0
-                }
-                val decodedImgDataResult = getMessageAsset(
-                    conversationId = message.conversationId,
-                    messageId = message.id
-                ).run {
-                    when (this) {
-                        is MessageAssetResult.Success -> decodedAsset
-                        else -> null
-                    }
-                }
-                when {
-                    content.value.mimeType.contains("image") -> ImageMessage(decodedImgDataResult, width = imgWidth, height = imgHeight)
-                    // TODO: To be changed once the error behavior has been defined with product
-                    assetId.isEmpty() -> TextMessage(messageBody = MessageBody("The asset message could not be downloaded correctly"))
-                    // TODO: Add generic asset message UI
-                    else -> TextMessage(MessageBody("GENERIC ASSET MESSAGE"))
-                }
-            }
+            is Asset -> mapToMessageUI(content.value, message.conversationId, message.id)
             is Text -> TextMessage(messageBody = MessageBody(content.value))
             else -> TextMessage(messageBody = MessageBody((content as? Text)?.value ?: "content is not available"))
         }
+
+    private suspend fun mapToMessageUI(assetContent: AssetContent, conversationId: ConversationId, messageId: String): MessageContent {
+        with(assetContent) {
+            val assetId = remoteData.assetId
+
+            // TODO: To be changed once the error behavior has been defined with product
+            if (assetId.isEmpty()) return TextMessage(messageBody = MessageBody("The asset message could not be downloaded correctly"))
+
+            val (imgWidth, imgHeight) = when (val md = metadata) {
+                is Image -> md.width to md.height
+                else -> 0 to 0
+            }
+            return when {
+                // If it's an image, we download it right away
+                mimeType.contains("image") -> ImageMessage(getRawAssetData(conversationId, messageId), width = imgWidth, height = imgHeight)
```

Comment:
what if user hides, or kills the app, while the image was downloading?

Reply:
Then the message will not be persisted, and next time user tries to display the message it will be redownloaded